subsys/net/lib/sockets/sockets_service.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_sock_svc, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/socket_service.h>
+#include <zephyr/posix/sys/eventfd.h>
+
+static int init_socket_service(void);
+static bool init_done;
+
+static K_MUTEX_DEFINE(lock);
+static K_CONDVAR_DEFINE(wait_start);
+
+static struct service {
+	/* The +1 is for triggering events from register function */
+	struct zsock_pollfd events[1 + CONFIG_NET_SOCKETS_POLL_MAX];
+	int count;
+} ctx;
+
+#define get_idx(svc) (*(svc->idx))
+
+void net_socket_service_foreach(net_socket_service_cb_t cb, void *user_data)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		cb(svc, user_data);
+	}
+}
+
+static void cleanup_svc_events(struct net_socket_service_desc *svc)
+{
+	for (int i = get_idx(svc); i < get_idx(svc) + svc->pev_len; i++) {
+		for (int j = 0; j < svc->pev_len; j++) {
+			if (ctx.events[i].fd >= 0 &&
+			    ctx.events[i].fd == svc->pev[j].event.fd) {
+				ctx.events[i].fd = -1;
+				svc->pev[j].event.fd = -1;
+				svc->pev[j].event.events = 0;
+			}
+		}
+	}
+}
+
+int z_impl_net_socket_service_register(const struct net_socket_service_desc *service,
+				       struct zsock_pollfd *fds, int len,
+				       void *user_data)
+{
+	int i, ret = -ENOENT;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	if (!init_done) {
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		if (svc == service) {
+			if (fds == NULL) {
+				cleanup_svc_events(svc);
+			} else {
+				if (len > svc->pev_len) {
+					NET_DBG("Too many file descriptors, "
+						"max is %d for service %p",
+						svc->pev_len, svc);
+					ret = -ENOMEM;
+					break;
+				}
+
+				for (i = 0; i < len; i++) {
+					svc->pev[i].event = fds[i];
+					svc->pev[i].user_data = user_data;
+				}
+
+				for (i = get_idx(svc); i < get_idx(svc) + svc->pev_len; i++) {
+					for (int j = 0; j < svc->pev_len; j++) {
+						ctx.events[i] = svc->pev[j].event;
+					}
+				}
+			}
+
+			/* Tell the thread to re-read the variables */
+			eventfd_write(ctx.events[0].fd, 1);
+			ret = 0;
+
+			break;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+static struct net_socket_service_desc *find_svc_and_event(
+	struct zsock_pollfd *pev,
+	struct net_socket_service_event **event)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		for (int i = 0; i < svc->pev_len; i++) {
+			if (svc->pev[i].event.fd == pev->fd) {
+				*event = &svc->pev[i];
+				return svc;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static void service_callback_sync(struct k_work *work)
+{
+	struct net_socket_service_event *pev =
+		CONTAINER_OF(work, struct net_socket_service_event, work);
+	struct net_socket_service_event ev = *pev;
+
+	ev.callback(&ev.work);
```

Comment:
What is happening here? Are you copying the event?
If so, why? And does this work with nested `work` objects?

Reply:
I did not want the callback to modify the event.
What nested work objects are you referring to?

Reply:
`ev.work`, or isn't that a `struct k_work` instance?

Reply:
Sure, but what kind of issue you see there. This work is not embedded to anything user has supplied and the API does not access the copied work.